File: Example/FWFramework/Classes/Module/Test/Module/TestPasscodeController.m
//
//  TestPasscodeController.m
//  FWFramework_Example
//
//  Created by wuyong on 2022/8/25.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

#import "TestPasscodeController.h"
#import "AppSwift.h"
@import FWFramework;

typedef NS_ENUM(NSInteger, CRBoxInputModelType) {
    CRBoxInputModelNormalType,
    CRBoxInputModelPlaceholderType,
    CRBoxInputModelCustomBoxType,
    CRBoxInputModelLineType,
    CRBoxInputModelSecretSymbolType,
    CRBoxInputModelSecretImageType,
    CRBoxInputModelSecretViewType,
};

@interface TestPasscodeController () <FWViewController>

@property (nonatomic, strong) NSMutableArray *dataArr;
@property (assign, nonatomic) CRBoxInputModelType boxInputModelType;

@property (nonatomic, strong) UIView *boxContainerView;
@property (nonatomic, strong) FWPasscodeView *boxInputView;
@property (nonatomic, strong) UIButton *securityButton;
@property (nonatomic, strong) UIButton *clearButton;
@property (strong, nonatomic) UILabel *valueLabel;

@end

@implementation TestPasscodeController

- (void)setupSubviews
{
    _valueLabel = [UILabel new];
    _valueLabel.textColor = AppTheme.textColor;
    _valueLabel.font = [UIFont boldSystemFontOfSize:24];
    _valueLabel.text = @"Empty";
    [self.view addSubview:_valueLabel];
    _valueLabel.fw_layoutChain.centerX().topToSafeAreaWithInset(30);
    
    _boxContainerView = [UIView new];
    [self.view addSubview:_boxContainerView];
    _boxContainerView.fw_layoutChain.leftWithInset(35).rightWithInset(35)
        .height(52).topToViewBottomWithOffset(_valueLabel, 30);
    
    _clearButton = [AppTheme largeButton];
    [_clearButton setTitle:@"Clear" forState:UIControlStateNormal];
    [_clearButton addTarget:self action:@selector(clearBtnEvent) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:_clearButton];
    _clearButton.fw_layoutChain.centerX().topToViewBottomWithOffset(_boxContainerView, 30);
    
    _securityButton = [AppTheme largeButton];
    [_securityButton setTitle:@"Security" forState:UIControlStateNormal];
    [_securityButton addTarget:self action:@selector(securityBtnEvent) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:_securityButton];
    _securityButton.fw_layoutChain.centerX().topToViewBottomWithOffset(_clearButton, 30);
}

- (void)setupNavbar
{
    FWWeakifySelf();
    [self fw_setRightBarItem:@"Toggle" block:^(id  _Nonnull sender) {
        FWStrongifySelf();
        NSArray *titles = [self.dataArr fw_mapWithBlock:^id _Nullable(NSArray *obj) {
            return obj.firstObject;
        }];
        [self fw_showSheetWithTitle:nil message:nil cancel:@"Cancel" actions:titles actionBlock:^(NSInteger index) {
            FWStrongifySelf();
            self.boxInputModelType = index;
        }];
    }];
}

- (void)setupLayout
{
    _dataArr = [NSMutableArray new];
    [_dataArr addObjectsFromArray:@[
        @[@"Normal", @(CRBoxInputModelNormalType)],
        @[@"Placeholder", @(CRBoxInputModelPlaceholderType)],
        @[@"Custom Box", @(CRBoxInputModelCustomBoxType)],
        @[@"Line", @(CRBoxInputModelLineType)],
        @[@"Secret Symbol", @(CRBoxInputModelSecretSymbolType)],
        @[@"Secret Image", @(CRBoxInputModelSecretImageType)],
        @[@"Secret View", @(CRBoxInputModelSecretViewType)],
    ]];
    
    self.boxInputModelType = CRBoxInputModelNormalType;
}

- (void)clearBtnEvent
{
    [_boxInputView clearAll];
}

- (void)securityBtnEvent
{
    _boxInputView.needSecurity = !_boxInputView.needSecurity;
}

#pragma mark - Setter & Getter

- (void)setBoxInputModelType:(CRBoxInputModelType)boxInputModelType
{
    _boxInputModelType = boxInputModelType;
    
    if (_boxInputView) {
        [_boxInputView removeFromSuperview];
        self.valueLabel.text = @"Empty";
    }
    
    switch (boxInputModelType) {
        case CRBoxInputModelNormalType:
            {
                _boxInputView = [self generateBoxInputView_normal];
            }
            break;
            
        case CRBoxInputModelPlaceholderType:
            {
                _boxInputView = [self generateBoxInputView_placeholder];
            }
            break;
            
        case CRBoxInputModelCustomBoxType:
            {
                _boxInputView = [self generateBoxInputView_customBox];
            }
            break;
            
        case CRBoxInputModelLineType:
            {
                _boxInputView = [self generateBoxInputView_line];
            }
            break;
            
        case CRBoxInputModelSecretSymbolType:
            {
                _boxInputView = [self generateBoxInputView_secretSymbol];
            }
            break;
            
        case CRBoxInputModelSecretImageType:
            {
                _boxInputView = [self generateBoxInputView_secretImage];
            }
            break;
            
        case CRBoxInputModelSecretViewType:
            {
                _boxInputView = [self generateBoxInputView_secretView];
            }
            break;
            
        default:
            {
                _boxInputView = [self generateBoxInputView_normal];
            }
            break;
    }
    
    __weak __typeof(self)weakSelf = self;
    if (!_boxInputView.textDidChangeBlock) {
        _boxInputView.textDidChangeBlock = ^(NSString *text, BOOL isFinished) {
            __strong __typeof(weakSelf)strongSelf = weakSelf;
            if (text.length > 0) {
                strongSelf.valueLabel.text = text;
            }else{
                strongSelf.valueLabel.text = @"Empty";
            }
        };
    }
    
    [self.boxContainerView addSubview:_boxInputView];
    _boxInputView.fw_layoutChain.edges();
}

#pragma mark - Normal
- (FWPasscodeView *)generateBoxInputView_normal
{
    FWPasscodeView *_boxInputView = [[FWPasscodeView alloc] initWithCodeLength:4];
    _boxInputView.collectionView.contentInset = UIEdgeInsetsMake(0, 20, 0, 20);
    _boxInputView.collectionView.contentOffset = CGPointMake(-40, 0);
    [_boxInputView prepareViewWithBeginEdit:YES];
    _boxInputView.inputType = FWPasscodeInputTypeNumber;
    
    _boxInputView.inputType = FWPasscodeInputTypeRegex;
    _boxInputView.customInputRegex = @"[^0-9]";
    
    if (@available(iOS 12.0, *)) {
        _boxInputView.textContentType = UITextContentTypeOneTimeCode;
    } else {
        _boxInputView.textContentType = @"one-time-code";
    }
    
    return _boxInputView;
}

- (FWPasscodeView *)generateBoxInputView_placeholder
{
    FWPasscodeCellProperty *cellProperty = [FWPasscodeCellProperty new];
    cellProperty.cellPlaceholderTextColor = [UIColor colorWithRed:114/255.0 green:116/255.0 blue:124/255.0 alpha:0.3];
    cellProperty.cellPlaceholderFont = [UIFont systemFontOfSize:20];
    
    FWPasscodeView *_boxInputView = [[FWPasscodeView alloc] initWithCodeLength:4];
    _boxInputView.collectionView.contentInset = UIEdgeInsetsMake(0, 20, 0, 20);
    _boxInputView.collectionView.contentOffset = CGPointMake(-40, 0);
    _boxInputView.showCursor = NO;
    _boxInputView.placeholderText = @"露可娜娜";
    _boxInputView.cellProperty = cellProperty;
    [_boxInputView prepareViewWithBeginEdit:YES];
    
    return _boxInputView;
}

#pragma mark - CustomBox
- (FWPasscodeView *)generateBoxInputView_customBox
{
    FWPasscodeCellProperty *cellProperty = [FWPasscodeCellProperty new];
    cellProperty.cellBgColorNormal = AppTheme.cellColor;
    cellProperty.cellBgColorSelected = [UIColor whiteColor];
    cellProperty.cellCursorColor = AppTheme.textColor;
    cellProperty.cellCursorWidth = 2;
    cellProperty.cellCursorHeight = (27);
    cellProperty.cornerRadius = 4;
    cellProperty.borderWidth = 0;
    cellProperty.cellFont = [UIFont boldSystemFontOfSize:24];
    cellProperty.cellTextColor = AppTheme.textColor;
    cellProperty.configCellShadowBlock = ^(CALayer * _Nonnull layer) {
        layer.shadowColor = [AppTheme.textColor colorWithAlphaComponent:0.2].CGColor;
        layer.shadowOpacity = 1;
        layer.shadowOffset = CGSizeMake(0, 2);
        layer.shadowRadius = 4;
    };

    FWPasscodeView *_boxInputView = [[FWPasscodeView alloc] initWithCodeLength:4];
    _boxInputView.collectionView.contentInset = UIEdgeInsetsMake(0, 10, 0, 10);
    _boxInputView.collectionView.contentOffset = CGPointMake(-20, 0);
    _boxInputView.flowLayout.itemSize = CGSizeMake((52), (52));
    _boxInputView.cellProperty = cellProperty;
    [_boxInputView prepareViewWithBeginEdit:YES];

    return _boxInputView;
}

#pragma mark - Line
- (FWPasscodeView *)generateBoxInputView_line
{
    FWPasscodeCellProperty *cellProperty = [FWPasscodeCellProperty new];
    cellProperty.cellCursorColor = AppTheme.textColor;
    cellProperty.cellCursorWidth = 2;
    cellProperty.cellCursorHeight = (27);
    cellProperty.cornerRadius = 0;
    cellProperty.borderWidth = 0;
    cellProperty.cellFont = [UIFont boldSystemFontOfSize:24];
    cellProperty.cellTextColor = AppTheme.textColor;
    cellProperty.showLine = YES;
    cellProperty.customLineViewBlock = ^FWPasscodeLineView * _Nonnull{
        FWPasscodeLineView *lineView = [FWPasscodeLineView new];
        lineView.underlineColorNormal = [AppTheme.textColor colorWithAlphaComponent:0.3];
        lineView.underlineColorSelected = [AppTheme.textColor colorWithAlphaComponent:0.7];
        lineView.underlineColorFilled = AppTheme.textColor;
        lineView.lineView.fw_layoutChain.remake().height(4).edgesWithInsetsExcludingEdge(UIEdgeInsetsZero, NSLayoutAttributeTop);
        
        lineView.selectChangeBlock = ^(FWPasscodeLineView * _Nonnull lineView, BOOL selected) {
            if (selected) {
                lineView.lineView.fw_layoutChain.height(6);
            } else {
                lineView.lineView.fw_layoutChain.height(4);
            }
        };

        return lineView;
    };

    FWPasscodeView *_boxInputView = [[FWPasscodeView alloc] initWithCodeLength:4];
    _boxInputView.collectionView.contentInset = UIEdgeInsetsMake(0, 10, 0, 10);
    _boxInputView.collectionView.contentOffset = CGPointMake(-20, 0);
    _boxInputView.flowLayout.itemSize = CGSizeMake((52), (52));
    _boxInputView.cellProperty = cellProperty;
    [_boxInputView prepareViewWithBeginEdit:YES];

    return _boxInputView;
}

#pragma mark - SecretSymbol
- (FWPasscodeView *)generateBoxInputView_secretSymbol
{
    FWPasscodeCellProperty *cellProperty = [FWPasscodeCellProperty new];
    cellProperty.cellCursorColor = AppTheme.textColor;
    cellProperty.cellCursorWidth = 2;
    cellProperty.cellCursorHeight = (27);
    cellProperty.cornerRadius = 0;
    cellProperty.borderWidth = 0;
    cellProperty.cellFont = [UIFont boldSystemFontOfSize:24];
    cellProperty.cellTextColor = AppTheme.textColor;
    cellProperty.showLine = YES;
    cellProperty.securitySymbol = @"*";//need

    FWPasscodeView *_boxInputView = [[FWPasscodeView alloc] initWithCodeLength:4];
    _boxInputView.collectionView.contentInset = UIEdgeInsetsMake(0, 10, 0, 10);
    _boxInputView.collectionView.contentOffset = CGPointMake(-20, 0);
    _boxInputView.needSecurity = YES;//need
    _boxInputView.flowLayout.itemSize = CGSizeMake((52), (52));
    _boxInputView.cellProperty = cellProperty;
    [_boxInputView prepareViewWithBeginEdit:NO];
    
    __weak __typeof(self)weakSelf = self;
    _boxInputView.textDidChangeBlock = ^(NSString *text, BOOL isFinished) {
        __strong __typeof(weakSelf)strongSelf = weakSelf;
        if (text.length > 0) {
            strongSelf.valueLabel.text = text;
        }else{
            strongSelf.valueLabel.text = @"Empty";
        }
    };
    
    _boxInputView.clearAllWhenEditingBegin = YES;
    [_boxInputView reloadInputString:@"5678"];
    
    return _boxInputView;
}

#pragma mark - SecretImage
- (FWPasscodeView *)generateBoxInputView_secretImage
{
    FWPasscodeCellProperty *cellProperty = [FWPasscodeCellProperty new];
    cellProperty.cellCursorColor = AppTheme.textColor;
    cellProperty.cellCursorWidth = 2;
    cellProperty.cellCursorHeight = (27);
    cellProperty.cornerRadius = 0;
    cellProperty.borderWidth = 0;
    cellProperty.cellFont = [UIFont boldSystemFontOfSize:24];
    cellProperty.cellTextColor = AppTheme.textColor;
    cellProperty.showLine = YES;
    cellProperty.securityType = FWPasscodeSecurityTypeView;//need
    cellProperty.customSecurityViewBlock = ^UIView * _Nonnull{
        FWPasscodeSecrectImageView *secrectImageView = [FWPasscodeSecrectImageView new];
        secrectImageView.image = FWIconImage(@"zmdi-var-settings", 24);
        secrectImageView.imageWidth = 23;
        secrectImageView.imageHeight = 23;
        return secrectImageView;
    };

    FWPasscodeView *_boxInputView = [[FWPasscodeView alloc] initWithCodeLength:4];
    _boxInputView.collectionView.contentInset = UIEdgeInsetsMake(0, 10, 0, 10);
    _boxInputView.collectionView.contentOffset = CGPointMake(-20, 0);
    _boxInputView.needSecurity = YES;//need
    _boxInputView.flowLayout.itemSize = CGSizeMake((52), (52));
    _boxInputView.cellProperty = cellProperty;
    [_boxInputView prepareViewWithBeginEdit:YES];

    return _boxInputView;
}

#pragma mark - SecretView
- (FWPasscodeView *)generateBoxInputView_secretView
{
    FWPasscodeCellProperty *cellProperty = [FWPasscodeCellProperty new];
    cellProperty.cellCursorColor = AppTheme.textColor;
    cellProperty.cellCursorWidth = 2;
    cellProperty.cellCursorHeight = (27);
    cellProperty.cornerRadius = 0;
    cellProperty.borderWidth = 0;
    cellProperty.cellFont = [UIFont boldSystemFontOfSize:24];
    cellProperty.cellTextColor = AppTheme.textColor;
    cellProperty.showLine = YES;
    cellProperty.securityType = FWPasscodeSecurityTypeView;//need
    cellProperty.customSecurityViewBlock = ^UIView * _Nonnull{
        UIView *customSecurityView = [UIView new];
        customSecurityView.backgroundColor = [UIColor clearColor];

        // circleView
        static CGFloat circleViewWidth = 20;
        UIView *circleView = [UIView new];
        circleView.backgroundColor = AppTheme.textColor;
        circleView.layer.cornerRadius = 4;
        [customSecurityView addSubview:circleView];
        circleView.fw_layoutChain.center().width(circleViewWidth).height(circleViewWidth);

        return customSecurityView;
    };

    FWPasscodeView *_boxInputView = [[FWPasscodeView alloc] initWithCodeLength:4];
    _boxInputView.collectionView.contentInset = UIEdgeInsetsMake(0, 10, 0, 10);
    _boxInputView.collectionView.contentOffset = CGPointMake(-20, 0);
    _boxInputView.needSecurity = YES;
    _boxInputView.flowLayout.itemSize = CGSizeMake((52), (52));
    _boxInputView.cellProperty = cellProperty;
    [_boxInputView prepareViewWithBeginEdit:YES];

    return _boxInputView;
}

@end

File: Example/FWFramework/Classes/Module/Test/Module/TestPasscodeController.swift
//
//  TestPasscodeController.swift
//  FWFramework_Example
//
//  Created by wuyong on 2022/10/18.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import FWFramework

enum TestPasscodeType: Int {
    case normal = 0
    case placeholder
    case custom
    case line
    case secretSymbol
    case secretImage
    case secretView
}

class TestPasscodeController: UIViewController, ViewControllerProtocol {
    
    var dataArray: [String] = [
        "Normal",
        "Placeholder",
        "Custom Box",
        "Line",
        "Secret Symbol",
        "Secret Image",
        "Secret View",
    ]
    
    var type: TestPasscodeType = .normal {
        didSet {
            updatePasscodeView()
        }
    }
    
    var boxInputView: PasscodeView?
    
    lazy var boxContainerView: UIView = {
        let result = UIView()
        return result
    }()
    
    lazy var securityButton: UIButton = {
        let result = AppTheme.largeButton()
        result.setTitle("Security", for: .normal)
        result.addTarget(self, action: #selector(securityButtonClicked), for: .touchUpInside)
        return result
    }()
    
    lazy var clearButton: UIButton = {
        let result = AppTheme.largeButton()
        result.setTitle("Clear", for: .normal)
        result.addTarget(self, action: #selector(clearButtonClicked), for: .touchUpInside)
        return result
    }()
    
    lazy var valueLabel: UILabel = {
        let result = UILabel()
        result.textColor = AppTheme.textColor
        result.font = UIFont.boldSystemFont(ofSize: 24)
        result.text = "Empty"
        return result
    }()
    
    func setupNavbar() {
        app.setRightBarItem(UIBarButtonItem.SystemItem.action) { [weak self] _ in
            self?.app.showSheet(title: nil, message: nil, actions: self?.dataArray, actionBlock: { index in
                self?.type = .init(rawValue: index) ?? .normal
            })
        }
    }
    
    func setupSubviews() {
        view.addSubview(valueLabel)
        view.addSubview(boxContainerView)
        view.addSubview(clearButton)
        view.addSubview(securityButton)
    }
    
    func setupLayout() {
        valueLabel.app.layoutChain
            .centerX()
            .top(toSafeArea: 30)
        
        boxContainerView.app.layoutChain
            .left(35)
            .right(35)
            .height(52)
            .top(toViewBottom: valueLabel, offset: 30)
        
        clearButton.app.layoutChain
            .centerX()
            .top(toViewBottom: boxContainerView, offset: 30)
        
        securityButton.app.layoutChain
            .centerX()
            .top(toViewBottom: clearButton, offset: 30)
        
        type = .normal
    }
    
    private func updatePasscodeView() {
        if let boxInputView = boxInputView {
            boxInputView.removeFromSuperview()
            valueLabel.text = "Empty"
        }
        
        var boxInputView: PasscodeView
        switch type {
        case .placeholder:
            boxInputView = generateBoxInputView_placeholder()
        case .custom:
            boxInputView = generateBoxInputView_custom()
        case .line:
            boxInputView = generateBoxInputView_line()
        case .secretSymbol:
            boxInputView = generateBoxInputView_secretSymbol()
        case .secretImage:
            boxInputView = generateBoxInputView_secretImage()
        case .secretView:
            boxInputView = generateBoxInputView_secretView()
        default:
            boxInputView = generateBoxInputView_normal()
        }
        self.boxInputView = boxInputView
        
        boxInputView.textDidChangeBlock = { [weak self] text, isFinished in
            if !text.isEmpty {
                self?.valueLabel.text = text
            } else {
                self?.valueLabel.text = "Empty"
            }
        }
        boxContainerView.addSubview(boxInputView)
        boxInputView.app.layoutChain.edges()
    }
    
    private func generateBoxInputView_normal() -> PasscodeView {
        let result = PasscodeView(codeLength: 4)
        result.collectionView.contentInset = UIEdgeInsets(top: 0, left: 20, bottom: 0, right: 20)
        result.collectionView.contentOffset = CGPoint(x: -40, y: 0)
        result.prepareView()
        result.inputType = .regex
        result.customInputRegex = "[^0-9]"
        result.textContentType = .oneTimeCode
        return result
    }
    
    private func generateBoxInputView_placeholder() -> PasscodeView {
        let cellProperty = PasscodeCellProperty()
        cellProperty.cellPlaceholderTextColor = UIColor(red: 114.0 / 255.0, green: 116.0 / 255.0, blue: 124.0 / 255.0, alpha: 0.3)
        cellProperty.cellPlaceholderFont = UIFont.systemFont(ofSize: 20)
        
        let result = PasscodeView(codeLength: 4)
        result.collectionView.contentInset = UIEdgeInsets(top: 0, left: 20, bottom: 0, right: 20)
        result.collectionView.contentOffset = CGPoint(x: -40, y: 0)
        result.showCursor = false
        result.placeholderText = "露可娜娜"
        result.cellProperty = cellProperty
        result.prepareView()
        return result
    }
    
    private func generateBoxInputView_custom() -> PasscodeView {
        let cellProperty = PasscodeCellProperty()
        cellProperty.cellBgColorNormal = AppTheme.cellColor
        cellProperty.cellBgColorSelected = .white
        cellProperty.cellCursorColor = AppTheme.textColor
        cellProperty.cellCursorWidth = 2
        cellProperty.cellCursorHeight = 27
        cellProperty.cornerRadius = 4
        cellProperty.borderWidth = 0
        cellProperty.cellFont = UIFont.boldSystemFont(ofSize: 24)
        cellProperty.cellTextColor = AppTheme.textColor
        cellProperty.configCellShadowBlock = { layer in
            layer.shadowColor = AppTheme.textColor.withAlphaComponent(0.2).cgColor
            layer.shadowOpacity = 1
            layer.shadowOffset = CGSize(width: 0, height: 2)
            layer.shadowRadius = 4
        }
        
        let result = PasscodeView(codeLength: 4)
        result.collectionView.contentInset = UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10)
        result.collectionView.contentOffset = CGPoint(x: -20, y: 0)
        result.flowLayout.itemSize = CGSize(width: 52, height: 52)
        result.cellProperty = cellProperty
        result.prepareView()
        return result
    }
    
    private func generateBoxInputView_line() -> PasscodeView {
        let cellProperty = PasscodeCellProperty()
        cellProperty.cellCursorColor = AppTheme.textColor
        cellProperty.cellCursorWidth = 2
        cellProperty.cellCursorHeight = 27
        cellProperty.cornerRadius = 0
        cellProperty.borderWidth = 0
        cellProperty.cellFont = UIFont.boldSystemFont(ofSize: 24)
        cellProperty.cellTextColor = AppTheme.textColor
        cellProperty.showLine = true
        cellProperty.customLineViewBlock = {
            let lineView = PasscodeLineView()
            lineView.underlineColorNormal = AppTheme.textColor.withAlphaComponent(0.3)
            lineView.underlineColorSelected = AppTheme.textColor.withAlphaComponent(0.7)
            lineView.underlineColorFilled = AppTheme.textColor
            lineView.lineView.app.layoutChain.remake()
                .height(4)
                .edges(excludingEdge: .top)
            lineView.selectChangeBlock = { lineView, selected in
                lineView.lineView.app.layoutChain.height(selected ? 6 : 4)
            }
            return lineView
        }
        
        let result = PasscodeView(codeLength: 4)
        result.collectionView.contentInset = UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10)
        result.collectionView.contentOffset = CGPoint(x: -20, y: 0)
        result.flowLayout.itemSize = CGSize(width: 52, height: 52)
        result.cellProperty = cellProperty
        result.prepareView()
        return result
    }
    
    private func generateBoxInputView_secretSymbol() -> PasscodeView {
        let cellProperty = PasscodeCellProperty()
        cellProperty.cellCursorColor = AppTheme.textColor
        cellProperty.cellCursorWidth = 2
        cellProperty.cellCursorHeight = 27
        cellProperty.cornerRadius = 0
        cellProperty.borderWidth = 0
        cellProperty.cellFont = UIFont.boldSystemFont(ofSize: 24)
        cellProperty.cellTextColor = AppTheme.textColor
        cellProperty.showLine = true
        cellProperty.securitySymbol = "*"
        
        let result = PasscodeView(codeLength: 4)
        result.collectionView.contentInset = UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10)
        result.collectionView.contentOffset = CGPoint(x: -20, y: 0)
        result.needSecurity = true
        result.flowLayout.itemSize = CGSize(width: 52, height: 52)
        result.cellProperty = cellProperty
        result.prepareView()
        
        result.clearAllWhenEditingBegin = true
        result.reloadInputString("5678")
        return result
    }
    
    private func generateBoxInputView_secretImage() -> PasscodeView {
        let cellProperty = PasscodeCellProperty()
        cellProperty.cellCursorColor = AppTheme.textColor
        cellProperty.cellCursorWidth = 2
        cellProperty.cellCursorHeight = 27
        cellProperty.cornerRadius = 0
        cellProperty.borderWidth = 0
        cellProperty.cellFont = UIFont.boldSystemFont(ofSize: 24)
        cellProperty.cellTextColor = AppTheme.textColor
        cellProperty.showLine = true
        cellProperty.securityType = .view
        cellProperty.customSecurityViewBlock = {
            let view = PasscodeSecrectImageView()
            view.image = APP.iconImage("zmdi-var-settings", 24) ?? UIImage()
            view.imageWidth = 23
            view.imageHeight = 23
            return view
        }
        
        let result = PasscodeView(codeLength: 4)
        result.collectionView.contentInset = UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10)
        result.collectionView.contentOffset = CGPoint(x: -20, y: 0)
        result.needSecurity = true
        result.flowLayout.itemSize = CGSize(width: 52, height: 52)
        result.cellProperty = cellProperty
        result.prepareView()
        return result
    }
    
    private func generateBoxInputView_secretView() -> PasscodeView {
        let cellProperty = PasscodeCellProperty()
        cellProperty.cellCursorColor = AppTheme.textColor
        cellProperty.cellCursorWidth = 2
        cellProperty.cellCursorHeight = 27
        cellProperty.cornerRadius = 0
        cellProperty.borderWidth = 0
        cellProperty.cellFont = UIFont.boldSystemFont(ofSize: 24)
        cellProperty.cellTextColor = AppTheme.textColor
        cellProperty.showLine = true
        cellProperty.securityType = .view
        cellProperty.customSecurityViewBlock = {
            let view = UIView()
            view.backgroundColor = .clear
            let circleView = UIView()
            circleView.backgroundColor = AppTheme.textColor
            circleView.layer.cornerRadius = 4
            view.addSubview(circleView)
            circleView.app.layoutChain.center()
                .width(20)
                .height(20)
            return view
        }
        
        let result = PasscodeView(codeLength: 4)
        result.collectionView.contentInset = UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10)
        result.collectionView.contentOffset = CGPoint(x: -20, y: 0)
        result.needSecurity = true
        result.flowLayout.itemSize = CGSize(width: 52, height: 52)
        result.cellProperty = cellProperty
        result.prepareView()
        return result
    }
    
    @objc func clearButtonClicked() {
        boxInputView?.clearAll()
    }
    
    @objc func securityButtonClicked() {
        guard let boxInputView = boxInputView else { return }
        boxInputView.needSecurity = !boxInputView.needSecurity
    }
    
}
